File: halma.js
// Copyright (c) 2013 by Jörg Müller <nexyon@gmail.com> All rights reserved.
// 
// This file is part of stern-halma.
// 
// stern-halma is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// stern-halma is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with stern-halma.  If not, see <http://www.gnu.org/licenses/>.

var radius = 24;
var drawradius = 6;

var EMPTY = 0;
var RED = 1;
var GREEN = 2;
var BLUE = 3;

var REGISTER = 0;
var LOGIN = 1;
var NEW_GAME = 2;
var JOIN_GAME = 3;
var CHANGE_GAME = 4;
var MOVE = 5;
var GAME_INFO = 6;
var TURN_INFO = 7;
var FIELD_INFO = 8;

var joinGameF;
var changeGameF;

window.onload = function () {
	var grab = {
		mx: 0,
		my: 0,
		drag: false,
		orig: {X: 0, Y:0},
		offsetx: 0,
		offsety: 0,
		pin: EMPTY
	};
	
	function drawCircle(ctx, pos, color) {
		ctx.beginPath();
		ctx.fillStyle = color;
		ctx.lineWidth = 2;
		ctx.arc(pos.X, pos.Y, drawradius, 0, 2 * Math.PI);
		ctx.fill();
		ctx.stroke();
	}
	
	function drawLine(ctx, p1, p2) {
		ctx.beginPath();
		ctx.lineWidth = 2;
		ctx.moveTo(p1.X, p1.Y);
		ctx.lineTo(p2.X, p2.Y);
		ctx.stroke();
	}
	
	function drawTriangle(ctx, p1, p2, p3, color) {
		p1 = gridToPixel(p1);
		p2 = gridToPixel(p2);
		p3 = gridToPixel(p3);
		
		ctx.beginPath();
		ctx.fillStyle = color;
		ctx.moveTo(p1.X, p1.Y);
		ctx.lineTo(p2.X, p2.Y);
		ctx.lineTo(p3.X, p3.Y);
		ctx.closePath();
		ctx.fill();
	}
	
	function dataToGrid(pos) {
		return { X: pos.X - 8, Y: pos.Y - 8 };
	}
	
	function gridToData(pos) {
		return { X: pos.X + 8, Y: pos.Y + 8 };
	}
	
	function gridToPixel(pos) {
		return { X: radius * Math.sqrt(3) * (pos.X + pos.Y / 2 + 6) + drawradius + 1,
		         Y: radius * 3 / 2 * (pos.Y + 8) + drawradius + 1 };
	}
	
	function pixelToGrid(pos) {
		var y = (pos.Y - drawradius - 1) * 2 / 3 / radius - 8;
		var x = (pos.X - drawradius - 1) / Math.sqrt(3) / radius - 6 - y / 2;
		var z = -x - y;
		
		var rx = Math.round(x);
		var ry = Math.round(y);
		var rz = Math.round(z);
		
		x = Math.abs(rx - x);
		y = Math.abs(ry - y);
		z = Math.abs(rz - z);
		
		if(x > y && x > z)
			rx = -ry - rz;
		else if(y > z)
			ry = -rx - rz;
		
		return { X: rx, Y: ry };
	}
	
	function drawCanvas() {
		var canvas = document.getElementById("board");
		var ctx = canvas.getContext("2d");
		ctx.clearRect(0, 0, canvas.width, canvas.height);
		
		drawTriangle(ctx, {X: 4, Y: -8}, {X: 0, Y: -4}, {X: 4, Y: -4}, "#FF0000");
		drawTriangle(ctx, {X: -4, Y: 4}, {X: 0, Y: 4}, {X: -4, Y: 8}, "#FF0000");
		drawTriangle(ctx, {X: -4, Y: -4}, {X: 0, Y: -4}, {X: -4, Y: 0}, "#00FF00");
		drawTriangle(ctx, {X: 4, Y: 0}, {X: 4, Y: 4}, {X: 0, Y: 4}, "#00FF00");
		drawTriangle(ctx, {X: 4, Y: -4}, {X: 8, Y: -4}, {X: 4, Y: 0}, "#0000FF");
		drawTriangle(ctx, {X: -4, Y: 0}, {X: -8, Y: 4}, {X: -4, Y: 4}, "#0000FF");
		
		for(var z = -7; z <= 7; z++) {
			var tz = Math.abs(z);
			
			var count = 8 + tz;
			var from = -4 - z;
			
			if(tz > 4) {
				count = 8 - tz;
			}
			
			if((tz > 4) ^ (z != tz)) {
				from = -4;
			}
			
			var p1 = gridToPixel({X: from, Y: z});
			var p2 = gridToPixel({X: from + count, Y: z});
			
			drawLine(ctx, p1, p2);
			
			p1 = gridToPixel({X: z, Y: -z - from});
			p2 = gridToPixel({X: z, Y: -z - from - count});
			
			drawLine(ctx, p1, p2);
			
			p1 = gridToPixel({X: -z - from, Y: from});
			p2 = gridToPixel({X: -z - from - count, Y: from + count});
			
			drawLine(ctx, p1, p2);
		}
		
		for(var u = -8; u <= 8; u++) {
			for(var v = -8; v <= 8; v++) {
				var tu = Math.abs(u);
				var tv = Math.abs(v);
				var tw = Math.abs(-u - v);
				
				if((tu <= 8 && tv <= 4 && tw <= 4) || (tu <= 4 && tv <= 8 && tw <=4) || (tu <= 4 && tv <= 4 && tw <=8)) {
					if((g = getField({X: u, Y: v})) != null) {
						if(g.pin == RED)
							drawCircle(ctx, gridToPixel({X: u, Y: v}), "#FF0000");
						else if(g.pin == GREEN)
							drawCircle(ctx, gridToPixel({X: u, Y: v}), "#00FF00");
						else if(g.pin == BLUE)
							drawCircle(ctx, gridToPixel({X: u, Y: v}), "#0000FF");
						/*else if((tu + tv + tw) / 2 < 4)
							drawCircle(ctx, gridToPixel({X: u, Y: v}), "#FFFF00");*/
						else
							drawCircle(ctx, gridToPixel({X: u, Y: v}), "#000000");
					}
				}
			}
		}
		
		for(var i = 0; i < possible.length; i++) {
			drawCircle(ctx, gridToPixel(possible[i]), "#FFFF00");
		}
		
		if(grab.drag) {
			var color = "#0000FF";
			if(grab.pin == RED)
				color = "#FF0000";
			else if(grab.pin == GREEN)
				color = "#00FF00";
			drawCircle(ctx, {X: grab.mx + grab.offsetx, Y: grab.my + grab.offsety}, color);
		}
	}
	
	function onMouseMove(event) {
		if(grab.drag) {
			var rect = canvas.getBoundingClientRect();
			
			pos = gridToPixel(pixelToGrid({X: event.clientX - rect.left, Y: event.clientY - rect.top}));
			grab.mx = event.clientX - rect.left;
			grab.my = event.clientY - rect.top;
			
			drawCanvas();
		}
	}
	
	function onMouseDown(event) {
		var rect = canvas.getBoundingClientRect();
		grab.mx = event.clientX - rect.left;
		grab.my = event.clientY - rect.top;
		pos = pixelToGrid({X: grab.mx, Y: grab.my});
		if(fieldValid(pos)) {
			pos = gridToData(pos);
			grab.orig = pos;
			if(fields[pos.X][pos.Y].pin != EMPTY) {
				grab.pin = fields[pos.X][pos.Y].pin;
				pos = gridToPixel(dataToGrid(pos));
				grab.offsetx = pos.X - grab.mx;
				grab.offsety = pos.Y - grab.my;
				if(grab.offsetx * grab.offsetx + grab.offsety * grab.offsety <= drawradius * drawradius) {
					fields[grab.orig.X][grab.orig.Y].pin = EMPTY;
					grab.drag = true;
					calculatePossible(dataToGrid(grab.orig));
					drawCanvas();
				}
			}
		}
	}
	
	function onMouseUp(event) {
		if(grab.drag) {
			grab.drag = false;
			possible = [];
			var worked = false;
			
			var rect = canvas.getBoundingClientRect();
			grab.mx = event.clientX - rect.left;
			grab.my = event.clientY - rect.top;
			pos = pixelToGrid({X: grab.mx + grab.offsetx, Y: grab.my + grab.offsety});
			
			if(fieldValid(pos) && current == me && grab.pin == me) {
				pos = gridToData(pos);
				if(fields[pos.X][pos.Y].pin == EMPTY) {
					//fields[pos.X][pos.Y].pin = grab.pin;
					fields[grab.orig.X][grab.orig.Y].pin = grab.pin;
					worked = true;
					move(grab.orig, pos);
				}
			}
			
			if(!worked)
				fields[grab.orig.X][grab.orig.Y].pin = grab.pin;
			
			drawCanvas();
		}
	}
	
	function getField(pos) {
		return fields[pos.X + 8][pos.Y + 8];
	}
	
	function setField(x, y, value) {
		fields[x + 8][y + 8] = value;
	}
	
	function createField(fields, x, y, type, pin) {
		fields[x + 8][y + 8] = {type: type, pin: pin};
	}
	
	function fieldValid(pos) {
		return Math.abs(pos.X) <= 8 && Math.abs(pos.Y) <= 8 && getField(pos) != null;
	}
	
	function calculatePossible(pos) {
		visited = [];
		for(var i = 0; i < 17; i++) {
			visited[i] = [];
			
			for(var j = 0; j < 17; j++)
				visited[i][j] = false;
		}
		
		var neighbors = [
			[+1,  0],  [+1, -1],  [ 0, -1],
			[-1,  0],  [-1, +1],  [ 0, +1] 
		];
				
		var todo = [pos];
		var first = true;

		while(todo.length > 0) {
			pos = todo.shift();
			if(visited[pos.X + 8][pos.Y + 8])
				continue;
			
			if(!first)
				possible.push(pos);
			visited[pos.X + 8][pos.Y + 8] = true;
			
			for(var i = 0; i < neighbors.length; i++) {
				var nb1 = {X: pos.X + neighbors[i][0], Y: pos.Y + neighbors[i][1]};
				var nb2 = {X: pos.X + 2 * neighbors[i][0], Y: pos.Y + 2 * neighbors[i][1]};
				if(fieldValid(nb1) && fieldValid(nb2) && getField(nb1).pin != EMPTY && getField(nb2).pin == EMPTY) {
					todo.push(nb2);
				}
				if(first && fieldValid(nb1) && getField(nb1).pin == EMPTY) {
					possible.push(nb1);
				}
			}
			
			first = false;
		}
	}
	
	function generateFields() {
		var fields = new Array(17);
		for(var i = 0; i < 17; i++)
			fields[i] = new Array(17);

		for(var u = -8; u <= 8; u++) {
			for(var v = -8; v <= 8; v++) {
				var tu = Math.abs(u);
				var tv = Math.abs(v);
				var tw = Math.abs(-u - v);
				
				if((tu <= 8 && tv <= 4 && tw <= 4) || (tu <= 4 && tv <= 8 && tw <=4) || (tu <= 4 && tv <= 4 && tw <=8)) {
					if((tu + tv + tw) / 2 < 4) {
						createField(fields, u, v, EMPTY, EMPTY);
					}
					else if(tv >= 4 && tv > tu && tv > tw) {
						if(v > 0)
							createField(fields, u, v, EMPTY, EMPTY);//RED);
						else
							createField(fields, u, v, RED, EMPTY);
					}
					else if(tw >= 4 && tw > tv && tw > tu) {
						if(-u - v > 0)
							createField(fields, u, v, EMPTY, EMPTY);//, GREEN);
						else
							createField(fields, u, v, GREEN, EMPTY);
					}
					else if(tu >= 4 && tu > tv && tu > tw) {
						if(u > 0)
							createField(fields, u, v, EMPTY, EMPTY);//, BLUE);
						else
							createField(fields, u, v, BLUE, EMPTY);
					}
				}
			}
		}
		
		createField(fields, 0, -4, RED, EMPTY);//, GREEN);
		createField(fields, -4, 0, BLUE, EMPTY);//, GREEN);
		createField(fields, -4, 4, BLUE, EMPTY);//, RED);
		createField(fields, 0, 4, GREEN, EMPTY);//, RED);
		createField(fields, 4, 0, GREEN, EMPTY);//, BLUE);
		createField(fields, 4, -4, RED, EMPTY);//, BLUE);
		
		return fields;
	}
	
	var fields = generateFields();
	
	var current = RED;
	
	var me = EMPTY;
	
	var possible = [];
	
	drawCanvas();
	
	canvas = document.getElementById("board");
	
	canvas.addEventListener("mousemove", onMouseMove);
	canvas.addEventListener("mousedown", onMouseDown);
	canvas.addEventListener("mouseup", onMouseUp);

	var connection = {
		connected: false,
		socket: null,
		
	};
	
	function colorToText(color) {
		switch(color) {
			case RED:
				return "red";
			case GREEN:
				return "green";
			case BLUE:
				return "blue";
			default:
				return "none";
		}
	}
	
	function onOpen(event) {
		console.log("Connected");
		connection.connected = true;
	}
	
	function onMessage(event) {
		message = JSON.parse(event.data);
		switch(message.Type) {
			case LOGIN:
				if(message.Player.OK) {
					document.getElementById("loginform").style.visibility = "hidden";
					sendMessage({Type: GAME_INFO});
					console.log("Login complete!");
				}
				else {
					// TODO: error message
					console.log("Login failed!");
				}
				break;
			case GAME_INFO:
				for(var i = 0; i < message.Games.length; i++) {
					var element = document.getElementById("game" + message.Games[i].ID);
					
					if(element == null) {
						element = document.createElement("tr");
						element.id = "game" + message.Games[i].ID;
						var table = document.getElementById("games");
						table.appendChild(element);
					}
					
					var newrow = "<td><a href=\"javascript:changeGameF(" + message.Games[i].ID + ");\">" + message.Games[i].ID + "</a></td><td>" + colorToText(message.Games[i].CurrentPlayer) + "</td><td>";
					
					if(message.Games[i].Player == EMPTY) {
						newrow += "<a href=\"javascript:joinGameF(" + message.Games[i].ID + ");\">" + colorToText(message.Games[i].Player) + "</a></td>";
					}
					else {
						newrow += colorToText(message.Games[i].Player) + "</td>";
					}
					
					element.innerHTML = newrow;
				}
				break;
			case CHANGE_GAME:
				sendMessage({Type: FIELD_INFO});
				me = message.Game.Player;
				document.getElementById("me").innerHTML = colorToText(me);
				current = message.Game.CurrentPlayer;
				document.getElementById("current").innerHTML = colorToText(current);
				break;
			case FIELD_INFO:
				for(var i = 0; i < message.Fields.length; i++) {
					pos = gridToData(message.Fields[i].Pos);
					fields[pos.X][pos.Y].pin = message.Fields[i].Pin;
				}
				drawCanvas();
				break;
			case TURN_INFO:
				current = message.Turn.CurrentPlayer;
				document.getElementById("current").innerHTML = colorToText(current);
				break;
			case MOVE:
				fields[message.Move.To.X + 8][message.Move.To.Y + 8].pin = fields[message.Move.From.X + 8][message.Move.From.Y + 8].pin;
				fields[message.Move.From.X + 8][message.Move.From.Y + 8].pin = EMPTY;
				drawCanvas();
				break;
		}
	}
	
	function onClose(event) {
		connection.connected = false;
		console.log("Closed");
	}
	
	function onError(event) {
		disconnect();
	}
	
	function connect() {
		connection.socket = new WebSocket("ws://localhost:8000/websocket/ws");
		connection.socket.onopen = onOpen;
		connection.socket.onmessage = onMessage;
		connection.socket.onerror = onError;
		connection.socket.onclose = onClose;
		console.log("Connecting");
	}

	function disconnect() {
		console.log("Disconnecting");
		connection.socket.close();
		connection.connected = false;
	}
	
	function sendMessage(message) {
		if(connection.connected) {
			connection.socket.send(JSON.stringify(message));
		}
	}
	
	function move(from, to) {
		sendMessage({Type: MOVE, Move: {From: dataToGrid(from), To: dataToGrid(to)}});
	}
	
	function register() {
		var name = document.getElementById("name").value;
		var password = document.getElementById("password").value;
		sendMessage({Type: REGISTER, Player: {Name: name, Password: password}});
	}
	
	function login() {
		var name = document.getElementById("name").value;
		var password = document.getElementById("password").value;
		sendMessage({Type: LOGIN, Player: {Name: name, Password: password}});
	}
	
	function newGame() {
		sendMessage({Type: NEW_GAME});
	}
	
	function refresh() {
		document.getElementById("games").innerHTML = "<tr><td>Game</td><td>Current Turn</td><td>My Color</td></tr>";
		sendMessage({Type: GAME_INFO});
	}
	
	function joinGame(id) {
		sendMessage({Type: JOIN_GAME, Game: {ID: id}});
	}
	
	function changeGame(id) {
		sendMessage({Type: CHANGE_GAME, Game: {ID: id}});
	}
	
	joinGameF = joinGame;
	changeGameF = changeGame;
	
	document.getElementById("register").onclick = register;
	document.getElementById("login").onclick = login;
	document.getElementById("newgame").onclick = newGame;
	document.getElementById("refresh").onclick = refresh;
	
	connect();
};

File: halma.go
// Copyright (c) 2013 by Jörg Müller <nexyon@gmail.com> All rights reserved.
// 
// This file is part of stern-halma.
// 
// stern-halma is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// stern-halma is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with stern-halma.  If not, see <http://www.gnu.org/licenses/>.

package main

import (
	"net/http"
	"fmt"
	"code.google.com/p/go.net/websocket"
	"math/rand"
)

type HalmaColor int

const (
	EMPTY HalmaColor = iota
	RED
	GREEN
	BLUE
)

type HalmaGameState int

const (
	NEW HalmaGameState = iota
	RUNNING
	DONE
)

type HalmaField struct {
	Type HalmaColor
	Pin HalmaColor
}

type Player struct {
	Name string
	Password string
}

type HalmaGame struct {
	ID int
	State HalmaGameState
	Fields [17][17]*HalmaField
	CurrentPlayer HalmaColor
	Players []*HalmaGamePlayer
	Clients []HalmaClient
}

type HalmaGamePlayer struct {
	Player *Player
	Game *HalmaGame
	Color HalmaColor
}

type HalmaClient interface {
	NotifyMove(color HalmaColor, from Position, to Position)
	NotifyTurn(current HalmaColor)
}

type Position struct {
	X int
	Y int
}

func abs(x int) int {
	if(x < 0) {
		return -x
	}
	return x
}

func (g *HalmaGame) createField(pos Position, Type HalmaColor, Pin HalmaColor) {
	g.Fields[pos.X + 8][pos.Y + 8] = &HalmaField{Type, Pin}
}

func (g *HalmaGame) fieldValid(pos Position) bool {
	return abs(pos.X) <= 8 && abs(pos.Y) <= 8 && g.getField(pos) != nil
}

func (g *HalmaGame) getField(pos Position) *HalmaField {
	return g.Fields[pos.X + 8][pos.Y + 8]
}

func (g *HalmaGame) calculatePossible(position Position) []Position {
	var visited [17][17]bool
	
	neighbors := []Position{
		{1, 0}, {1, -1}, {0, -1},
		{-1, 0}, {-1, 1}, {0, 1},
	}
	
	todo := []Position{position}
	possible := make([]Position, 0)
	first := true
	
	var pos Position

	for len(todo) > 0 {
		pos, todo = todo[len(todo) - 1], todo[:len(todo) - 1]
		if visited[pos.X + 8][pos.Y + 8] {
			continue
		}
		
		if !first {
			possible = append(possible, pos)
		}
		visited[pos.X + 8][pos.Y + 8] = true
		
		for i := 0; i < len(neighbors); i++ {
			nb1 := Position{pos.X + neighbors[i].X, pos.Y + neighbors[i].Y}
			nb2 := Position{pos.X + 2 * neighbors[i].X, pos.Y + 2 * neighbors[i].Y}
			if g.fieldValid(nb1) && g.fieldValid(nb2) && g.getField(nb1).Pin != EMPTY && g.getField(nb2).Pin == EMPTY {
				todo = append(todo, nb2)
			}
			if first && g.fieldValid(nb1) && g.getField(nb1).Pin == EMPTY {
				possible = append(possible, nb1)
			}
		}
		
		first = false
	}
	
	return possible
}

func (g *HalmaGame) checkWinner(color HalmaColor) bool {
	for x := 0; x < len(g.Fields); x++ {
		for y := 0; y < len(g.Fields[x]); y++ {
			if g.Fields[x][y] != nil {
				if g.Fields[x][y].Type == color && g.Fields[x][y].Type != g.Fields[x][y].Pin {
					return false
				}
			}
		}
	}
	
	return true
}

func (g *HalmaGame) GetFreeColor() HalmaColor {
	color := HalmaColor(len(g.Players) + 1)
	if color > 3 {
		color = EMPTY
	}
	return color
}

func (g *HalmaGame) GetPlayerColor(player *Player) HalmaColor {
	hgp := g.GetPlayer(player)

	if hgp != nil {
		return hgp.Color
	}

	return EMPTY
}

func NewHalmaGame() *HalmaGame {
	g := &HalmaGame{len(gameList), NEW, [17][17]*HalmaField{}, EMPTY, nil, nil}
	gameList = append(gameList, g)

	return g
}

func (g *HalmaGame) Start() {
	green := EMPTY
	blue := EMPTY
	
	if len(g.Players) > 1 {
		green = GREEN
		
		if len(g.Players) > 2 {
			blue = BLUE
		}
	}
	
	for u := -8; u <= 8; u++ {
		for v := -8; v <= 8; v++ {
			tu := abs(u)
			tv := abs(v)
			tw := abs(-u - v)
			
			if (tu <= 8 && tv <= 4 && tw <= 4) || (tu <= 4 && tv <= 8 && tw <=4) || (tu <= 4 && tv <= 4 && tw <=8) {
				if (tu + tv + tw) / 2 < 4 {
					g.createField(Position{u, v}, EMPTY, EMPTY)
				} else if tv >= 4 && tv > tu && tv > tw  {
					if v > 0 {
						g.createField(Position{u, v}, EMPTY, RED)
					} else {
						g.createField(Position{u, v}, RED, EMPTY)
					}
				} else if tw >= 4 && tw > tv && tw > tu {
					if -u - v > 0 {
						g.createField(Position{u, v}, EMPTY, green)
					} else {
						g.createField(Position{u, v}, GREEN, EMPTY)
					}
				} else if tu >= 4 && tu > tv && tu > tw {
					if u > 0 {
						g.createField(Position{u, v}, EMPTY, blue)
					} else {
						g.createField(Position{u, v}, BLUE, EMPTY)
					}
				}
			}
		}
	}
	
	g.createField(Position{0, -4}, RED, green)
	g.createField(Position{-4, 0}, BLUE, green)
	g.createField(Position{-4, 4}, BLUE, RED)
	g.createField(Position{0, 4}, GREEN, RED)
	g.createField(Position{4, 0}, GREEN, blue)
	g.createField(Position{4, -4}, RED, blue)
	
	g.State = RUNNING
	g.CurrentPlayer = HalmaColor(rand.Intn(len(g.Players)) + 1)
}

type HalmaMessageType int

const (
	REGISTER HalmaMessageType = iota
	LOGIN
	NEW_GAME
	JOIN_GAME
	CHANGE_GAME
	MOVE
	GAME_INFO
	TURN_INFO
	FIELD_INFO
)

type PlayerMessage struct {
	Name string
	Password string
	OK bool
}

type TurnMessage struct {
	CurrentPlayer HalmaColor
}

type MoveMessage struct {
	From Position
	To Position
}

type HalmaMessage struct {
	Type HalmaMessageType
	Player *PlayerMessage
	Fields []FieldInfo
	Games []GameInfo
	Turn *TurnMessage
	Move *MoveMessage
	Game *GameInfo
}

type FieldInfo struct {
	Pos Position
	Pin HalmaColor
}

type GameInfo struct {
	ID int
	Player HalmaColor
	CurrentPlayer HalmaColor
}

func (g *HalmaGame) Move(color HalmaColor, from Position, to Position) {
	if color != g.CurrentPlayer || !g.fieldValid(from) || !g.fieldValid(to) || g.getField(from).Pin != color {
		return
	}

	possible := g.calculatePossible(from)
	
	for _, pos := range(possible) {
		if pos.X == to.X && pos.Y == to.Y {
			g.Fields[to.X + 8][to.Y + 8].Pin = g.Fields[from.X + 8][from.Y + 8].Pin
			g.Fields[from.X + 8][from.Y + 8].Pin = EMPTY
			
			if g.checkWinner(g.CurrentPlayer) {
				g.CurrentPlayer = EMPTY
			} else {
				g.CurrentPlayer++
				if int(g.CurrentPlayer) > len(g.Players) {
					g.CurrentPlayer = RED
				}
			}
			
			for _, client := range(g.Clients) {
				client.NotifyMove(color, from, to)
				client.NotifyTurn(g.CurrentPlayer)
			}
		}
	}
}

func (g *HalmaGame) Join(player *Player) *HalmaGamePlayer {
	color := g.GetFreeColor()
	
	if color == EMPTY {
		return nil
	}
	
	hgp := &HalmaGamePlayer{player, g, color}
	g.Players = append(g.Players, hgp)
	
	return hgp
}

func (g *HalmaGame) GetPlayer(player *Player) *HalmaGamePlayer {
	for _, hgp := range(g.Players) {
		if hgp.Player == player {
			return hgp
		}
	}
	
	return nil
}

func NewPlayer(name string, password string) *Player {
	return &Player{name, password}
}

func DestroyClient(client *HalmaWebsocketClient) {
	if client != nil && client.Player != nil {
		game := client.Player.Game
		for i := range(game.Clients) {
			if game.Clients[i] == client {
				game.Clients[i], game.Clients = game.Clients[len(game.Clients)-1], game.Clients[:len(game.Clients)-1]
				break
			}
		}
	}
}

func process(ws *websocket.Conn) {
	var player *Player = nil
	var client *HalmaWebsocketClient = nil
	var game *HalmaGame = nil
	
	//fmt.Println("Someone connected!")
	
	for {
		var message HalmaMessage
		err := websocket.JSON.Receive(ws, &message)
		
		if err != nil {
			fmt.Println("Error: ", err)
			break
		}
		
		//fmt.Println("Received message!")
		
		switch(message.Type) {
			case MOVE:
				if client != nil && client.Player != nil {
					game.Move(client.Player.Color, message.Move.From, message.Move.To)
				}
				break
			case FIELD_INFO:
				var response HalmaMessage
				response.Type = FIELD_INFO
				response.Fields = make([]FieldInfo, 0)
				for x := 0; x < len(game.Fields); x++ {
					for y := 0; y < len(game.Fields[x]); y++ {
						if game.Fields[x][y] != nil {
							response.Fields = append(response.Fields, FieldInfo{Position{x - 8, y - 8}, game.Fields[x][y].Pin})
						}
					}
				}
				websocket.JSON.Send(ws, &response)
				break
			case REGISTER:
				if message.Player != nil {
					player = RegisterPlayer(message.Player.Name, message.Player.Password)
					var response HalmaMessage
					response.Type = LOGIN
					response.Player = &PlayerMessage{message.Player.Name, "", player != nil}
					websocket.JSON.Send(ws, &response)
				}
				break
			case LOGIN:
				if message.Player != nil {
					player = LoginPlayer(message.Player.Name, message.Player.Password)
					var response HalmaMessage
					response.Type = LOGIN
					response.Player = &PlayerMessage{message.Player.Name, "", player != nil}
					websocket.JSON.Send(ws, &response)
				}
				break
			case GAME_INFO:
				var response HalmaMessage
				response.Type = GAME_INFO
				response.Games = make([]GameInfo, len(gameList))
				for i, game := range(gameList) {
					response.Games[i].ID = game.ID
					response.Games[i].Player = game.GetPlayerColor(player)
					response.Games[i].CurrentPlayer = game.CurrentPlayer
				}
				websocket.JSON.Send(ws, &response)
				break
			case CHANGE_GAME:
				if message.Game == nil || message.Game.ID < 0 || message.Game.ID >= len(gameList) {
					break
				}
				
				game = gameList[message.Game.ID]
				
				hgp := game.GetPlayer(player)
				
				DestroyClient(client)
				client = &HalmaWebsocketClient{ws, hgp}
				game.Clients = append(game.Clients, client)
				
				var response HalmaMessage
				response.Type = CHANGE_GAME
				if hgp == nil {
					response.Game = &GameInfo{game.ID, EMPTY, game.CurrentPlayer}
				} else {
					if game.State == NEW {
						game.Start()
					}
					
					response.Game = &GameInfo{game.ID, hgp.Color, game.CurrentPlayer}
				}
				websocket.JSON.Send(ws, &response)
				break
			case TURN_INFO:
				var response HalmaMessage
				response.Type = TURN_INFO
				response.Turn = &TurnMessage{game.CurrentPlayer}
				websocket.JSON.Send(ws, &response)
				break
			case NEW_GAME:
				game := NewHalmaGame()
				game.Join(player)
				
				var response HalmaMessage
				response.Type = GAME_INFO
				response.Games = []GameInfo{GameInfo{game.ID, game.GetPlayerColor(player), game.CurrentPlayer}}
				websocket.JSON.Send(ws, &response)
				break
			case JOIN_GAME:
				if message.Game == nil || message.Game.ID < 0 || message.Game.ID >= len(gameList) {
					break
				}
				
				game := gameList[message.Game.ID]
				
				if game.State != NEW {
					break
				}
				
				hgp := game.Join(player)
				
				var response HalmaMessage
				response.Type = GAME_INFO
				response.Games = make([]GameInfo, 1)
				response.Games[0].ID = game.ID
				response.Games[0].Player = hgp.Color
				response.Games[0].CurrentPlayer = game.CurrentPlayer
				websocket.JSON.Send(ws, &response)
				break
		}
	}
	
	DestroyClient(client)
	
	fmt.Println("Closing socket")
}

func RegisterPlayer(name string, password string) *Player {
	for _, player := range(playerList) {
		if player.Name == name {
			return nil
		}
	}
	
	player := &Player{name, password}
	playerList = append(playerList, player)
	return player
}

func LoginPlayer(name string, password string) *Player {
	for _, player := range(playerList) {
		if player.Name == name && player.Password == password {
			return player
		}
	}

	return nil
}

type HalmaWebsocketClient struct {
	Connection *websocket.Conn
	Player *HalmaGamePlayer
}

func (c *HalmaWebsocketClient) NotifyMove(color HalmaColor, from Position, to Position) {
	var message HalmaMessage
	
	message.Type = FIELD_INFO
	message.Fields = make([]FieldInfo, 2)
	message.Fields[0].Pos.X = from.X
	message.Fields[0].Pos.Y = from.Y
	message.Fields[0].Pin = EMPTY
	message.Fields[1].Pos.X = to.X
	message.Fields[1].Pos.Y = to.Y
	message.Fields[1].Pin = color

	websocket.JSON.Send(c.Connection, &message)
}

func (c *HalmaWebsocketClient) NotifyTurn(current HalmaColor) {
	var message HalmaMessage
	message.Type = TURN_INFO
	message.Turn = &TurnMessage{current}
	websocket.JSON.Send(c.Connection, &message)
}

var gameList []*HalmaGame
var playerList []*Player

func main() {
	gameList = make([]*HalmaGame, 0)
	playerList = make([]*Player, 0)
	
	http.Handle("/websocket/", websocket.Handler(process))
	http.Handle("/", http.FileServer(http.Dir(".")))
	err := http.ListenAndServe(":8000", nil)
	
	if err != nil {
		fmt.Println(err)
	}
}
